File: backend/app/schemas/system.py
"""Pydantic models for system probe responses."""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class OSInfo(BaseModel):
    system: str
    release: str
    version: str
    machine: str


class CPUInfo(BaseModel):
    model: Optional[str]
    architecture: Optional[str]
    cores_physical: Optional[int]
    cores_logical: Optional[int]
    max_frequency_mhz: Optional[float]


class MemoryInfo(BaseModel):
    total_gb: Optional[float]
    available_gb: Optional[float]


class GPUDevice(BaseModel):
    name: Optional[str]
    memory_gb: Optional[float]
    multi_processor_count: Optional[int]


class GPUInfo(BaseModel):
    has_gpu: bool
    api: Optional[str]
    driver: Optional[str]
    devices: List[GPUDevice]


class DiskUsage(BaseModel):
    path: str
    total_gb: Optional[float]
    used_gb: Optional[float]
    free_gb: Optional[float]


class StorageInfo(BaseModel):
    database: Optional[DiskUsage] = None
    media: DiskUsage
    transcripts: DiskUsage
    project: DiskUsage


class NetworkInterface(BaseModel):
    name: str
    ipv4: List[str]


class NetworkInfo(BaseModel):
    hostname: str
    interfaces: List[NetworkInterface]


class RuntimeInfo(BaseModel):
    python: str
    node: Optional[str]


class ContainerInfo(BaseModel):
    is_container: bool = Field(..., alias="is_container")
    indicators: List[str]

    model_config = {"populate_by_name": True}


class RecommendationInfo(BaseModel):
    suggested_asr_model: str
    suggested_diarization: str
    basis: List[str]


class SystemProbeResponse(BaseModel):
    detected_at: datetime
    os: OSInfo
    cpu: CPUInfo
    memory: MemoryInfo
    gpu: GPUInfo
    storage: StorageInfo
    network: NetworkInfo
    runtime: RuntimeInfo
    container: ContainerInfo
    recommendation: RecommendationInfo

File: frontend/src/services/system.ts
import { apiGet, apiPost } from '../lib/api';

export interface GPUDevice {
  name?: string;
  memory_gb?: number;
  multi_processor_count?: number;
}

export interface SystemProbe {
  detected_at: string;
  os: {
    system: string;
    release: string;
    version: string;
    machine: string;
  };
  cpu: {
    model?: string;
    architecture?: string;
    cores_physical?: number;
    cores_logical?: number;
    max_frequency_mhz?: number;
  };
  memory: {
    total_gb?: number;
    available_gb?: number;
  };
  gpu: {
    has_gpu: boolean;
    api?: string | null;
    driver?: string | null;
    devices: GPUDevice[];
  };
  storage: {
    database?: DiskUsage | null;
    media: DiskUsage;
    transcripts: DiskUsage;
    project: DiskUsage;
  };
  network: {
    hostname: string;
    interfaces: Array<{ name: string; ipv4: string[] }>;
  };
  runtime: {
    python: string;
    node?: string | null;
  };
  container: {
    is_container: boolean;
    indicators: string[];
  };
  recommendation: {
    suggested_asr_model: string;
    suggested_diarization: string;
    basis: string[];
  };
}

export interface DiskUsage {
  path: string;
  total_gb?: number | null;
  used_gb?: number | null;
  free_gb?: number | null;
}

export async function fetchSystemInfo(): Promise<SystemProbe> {
  return apiGet<SystemProbe>('/system/info');
}

export async function refreshSystemInfo(): Promise<SystemProbe> {
  return apiPost<SystemProbe>('/system/info/detect', {});
}

export interface CapabilityResponse {
  asr: Array<{
    provider: string;
    display_name: string;
    available: boolean;
    models: string[];
    notes: string[];
  }>;
  diarizers: Array<{
    key: string;
    display_name: string;
    requires_gpu: boolean;
    available: boolean;
    notes: string[];
  }>;
}

export async function fetchCapabilities(): Promise<CapabilityResponse> {
  return apiGet<CapabilityResponse>('/system/availability');
}
